### PR TITLE
docs: add notebook for openai finetuning with kili

### PR DESCRIPTION
**Pull request by Jonas1312**

### CODE REVIEW

File: recipes/Kili_fine_tuning_tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,2081 @@
+{
```
Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>it's a bit weird because in the first paragraph you already present the dataset we'll use, and then you say "we'll start by finding..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>I think that here you should explain that each line of this file is an input text and its class, separated by an "|"</p><p> </p><p>Also, I would change the code like this instead:</p><p> </p><pre class="ql-syntax" spellcheck="false">
with open(...) as f:
	lines = f.readlines()
 
lines = [line.split("|") for line in lines]
content_array = [line[0] for line in lines]
categories_array = [line[1] for line in lines]
external_id_array = [f"text_{i}" for i, _ in enumerate(content_array)]
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>I don't think the organization id is needed, not sure...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>you can say that as of now, only GPT3 models can be finetuned : https://platform.openai.com/docs/models/gpt-3</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.    text: """
</pre><p>why "text:" and not "predicted class:" for example?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #12.        max_tokens=8,
</pre><p>you should explain how you determine the max number of tokens to generate</p><p> </p><p>by using the openai tokenizer for example: https://platform.openai.com/tokenizer</p><p> </p><p>the longest class has 7 tokens, with some margin (since the LLM can add spaces somtimes), it's best to fix it to 10 or 15</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">&nbsp; index = len(returned_value)
&nbsp; for predefined_class in all_predefined_classes:
&nbsp;&nbsp;&nbsp; if predefined_class in returned_value.upper() and returned_value.upper().index(predefined_class) &lt; index:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; result = predefined_class
</pre><p> </p><p>so in this part I recommend you to use a more advanced parser &amp; sanitizer for cleaning the openai prediction (you have to adapt the code to your notebook):</p><p> </p><pre class="ql-syntax" spellcheck="false">def parse_and_sanitize_openai_response(response: str, categories) -&gt; str:
&nbsp;&nbsp;&nbsp; if response in categories:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; return response

&nbsp;&nbsp;&nbsp; response = response.replace("OUTPUT:", "")
&nbsp;&nbsp;&nbsp; if response in categories:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; return response

&nbsp;&nbsp;&nbsp; response = response.strip()&nbsp; # remove spaces at the start/end
&nbsp;&nbsp;&nbsp; if response in categories:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; return response

&nbsp;&nbsp;&nbsp; response = response.strip(".")
&nbsp;&nbsp;&nbsp; if response in categories:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; return response

&nbsp;&nbsp;&nbsp; min_distance = 1e9
&nbsp;&nbsp;&nbsp; closest_category = None
&nbsp;&nbsp;&nbsp; for class_name in categories:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; dist = Levenshtein.distance(class_name, response)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; if dist &lt; min_distance:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; closest_category = class_name

&nbsp;&nbsp;&nbsp; return closest_category
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>uncomment?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">y_pred = []

for content in content_array:
	...
	y_pred.append(parsed_and_sanitized_class(predicted_class))
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    inference_cat_array = ["UNABLE_TO_CLASSIFY"] * 800
</pre><p>where does the 800 come from?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>you can say here that inference_mark_lte is used to filter assets that have an agreement score lower than 51%</p><p> </p><p>you could use 50% directly, since lte stands for "lower than or equal to"</p><p> </p><p>you can say that 671 examples have been wrongly classified, that is the model was only right 16% of the time</p><p> </p><p>we have 8 classes, so a random "dumb" model would be able to have 1/8 = 12.5% accuracy.</p><p> </p><p>So our LLM is a little bit smarter than a dumb model</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>we usually assume that score close to 1 is a good score</p><p> </p><p>so here, I would use (800-671)/800 to get the score instead</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>as high</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>you can also retrieve the assets and labels directly using kili.assets or kili.labels, so that you skip the unzip and json loading/parsing step</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>correct link: https://python-sdk-docs.kili-technology.com/2.135/sdk/label/#kili.entrypoints.queries.label.__init__.QueriesLabel.export_labels</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #22.        exported_class = (nested_field[0]['name'])
</pre><p>I would just use exported_class = json_data["latestLabel"]["jsonResponse"]['CLASSIFICATION_JOB']['categories'][0]["name"]</p><p> </p><p>or use the label parsing</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='46'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    sample_text2 = "Most Of Beijing To Be Tested For COVID-19 Amid Lockdown Worry. While only 70 cases have been found since the outbreak surfaced, authorities have followed a “zero-COVID” approach to try to prevent a further spread of the virus."
</pre><p>remove this line?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='58'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>explain that we start at 801 to assign new external ids</p><p> </p><p>you can use the same code I gave above in the notebook, not dict needed for that</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='60'/>

Comment (Jonas1312) [inline recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>I would have expected better score after finetuning, weird</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kili-technology/kili-python-sdk/pull/1259/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='70'/>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>The dataset is not mentioned earlier in the NB.</p>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>1: addressed in NB</p><p>2. Will address this in version 2.0 of the NB</p>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>This is based on what I found in OpenAI's documentation.</p>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>Done</p>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>Will update this in version 2.0 of this NB</p>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>Will update this in version 2.0 of the NB</p>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>I'm creating a list with fixed length and want to be able to control where the results will get appended based on a dataset that's 800 records long. So that I don't end up with a mismatch. I'll add n explanation that 800 is based on the dataset length</p>

[the remark this answers is not in the document]
Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>ok, removed</p>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>I've made this change. Closing</p>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>Agreed. I've made this change</p>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>True. I'll make this change in version 2.0 of the NB</p>

[the remark this answers is not in the document]
Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>Done. Added a line with !head /content/kili-fine-tune.jsonl</p>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>Commented out. This is a sample line in case if someone wanted to experiment a bit more.</p>

Comment (xczuba) [reply recipes/Kili_fine_tuning_tutorial.ipynb:1]: <p>II agree. Maybe I'll run it by Jean L, just in case.</p>
